### PR TITLE
61 improve halo2 query calls

### PR DESCRIPTION
**Pull request by CeciliaZ030**
[#61](https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1115#issuecomment-1415986963)

### CODE REVIEW

File: halo2_proofs/src/plonk/circuit.rs [hunk 1]
```diff
@@ -19,21 +23,25 @@ mod compress_selectors;
...
+'static + Sized + Copy + std::fmt::Debug + PartialEq + Eq + Into<Any>
 {
+    /// Return expression from cell
+    fn query_cell<F: Field>(&self, index: usize, at: Rotation) -> Expression<F>;
 }
 
+
 /// A column with an index and type
-#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:34]: `Column` seems to have not been changed, so is this change still necessary?

File: halo2_proofs/src/plonk/circuit.rs [hunk 2]
```diff
@@ -45,9 +53,36 @@ impl<C: ColumnType> Column<C> {
...
+    }
 }
 
-impl<C: ColumnType> Ord for Column<C> {
+
+impl<C: ColumnType> PartialOrd<Self> for Column<C> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:83]: Same here as above, still necessary?

File: halo2_proofs/src/plonk/circuit.rs [hunk 3]
```diff
@@ -166,6 +203,10 @@ impl std::fmt::Debug for Advice {
 #[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
 pub struct Fixed;
 
+impl Fixed {
+
+}
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:208]: Can be removed I assume.

File: halo2_proofs/src/plonk/circuit.rs [hunk 4]
```diff
@@ -354,7 +442,7 @@ impl TryFrom<Column<Any>> for Column<Instance> {
 /// row when required:
 /// ```
 /// use halo2_proofs::{
-///     arithmetic::FieldExt,
+///     arithmetic::Field,
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:445]: I think best to revert these changes to keep the diff to a minimum.

File: halo2_proofs/src/plonk/circuit.rs [hunk 5]
```diff
@@ -705,7 +808,44 @@ pub enum Expression<F> {
     Scaled(Box<Expression<F>>, F),
 }
 
+
+
 impl<F: Field> Expression<F> {
+
+    /// Make side effects
+    pub fn load_to_virtual_cells(&mut self, cells: &mut VirtualCells<'_, F>) {
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:816]:
```suggestion
    /// Queries all column cells
    pub fn query_cells(&mut self, cells: &mut VirtualCells<'_, F>) {
```
Or something like that.

File: halo2_proofs/src/plonk/circuit.rs [hunk 6]
```diff
@@ -705,7 +808,44 @@ pub enum Expression<F> {
...
+            Expression::Advice(query) => {
+                let col = Column{ index: query.column_index, column_type: Advice{phase: query.phase}};
+                cells.queried_cells.push((col, query.rotation).into());
+                query.index = Some(cells.meta.query_advice_index(col, query.rotation));
+            }
+            Expression::Instance(query) => {
+                let col = Column{ index: query.column_index, column_type: Instance};
+                cells.queried_cells.push((col, query.rotation).into());
+                query.index = Some(cells.meta.query_instance_index(col, query.rotation));
+            },
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:834]: Is it possible to simply reuse `query_fixed`, `query_advice` and `query_instance` here? Because now the query index is set in two different places using two versions of the code which, admittedly, is currently very short, but still I think it would be great to only do it in one place to make sure the same logic is always done no matter how the cell is queried. I would think something like `self = cells.query_advice(...)` could work.

File: halo2_proofs/src/plonk/circuit.rs [hunk 7]
```diff
@@ -1208,6 +1348,34 @@ impl<Col: Into<Column<Any>>> From<(Col, Rotation)> for VirtualCell {
...
+// impl From<InstanceQuery> for VirtualCell {
+//     fn from(query: InstanceQuery) -> VirtualCell{
+//         VirtualCell{
+//             column: Column {index: query.column_index, column_type: Any::from(Instance) },
+//             rotation: query.rotation
+//         }
+//     }
+// }
+
+
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:1378]: remove?

File: halo2_proofs/src/plonk/circuit.rs [hunk 8]
```diff
@@ -1696,15 +1864,18 @@ impl<F: Field> ConstraintSystem<F> {
...
-            .map(|c| (c.name, c.poly))
+            .map(|mut c: Constraint<F>| {
+                c.poly.load_to_virtual_cells(&mut cells);
+                (c.name, c.poly)
+            })
             .unzip();
 
+        let queried_selectors = cells.queried_selectors;
+        let queried_cells = cells.queried_cells;
+
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:1878]: Does this need to move here from its original position?

**Review by Brechtpd — COMMENTED** [review]
Looks great! Added some suggestions, also:
- The diff contains some whitespace/format changes which I think will largely be solved by running `cargo fmt` (code won't be accepted in the PSE repo without it). Aiming for minimal diffs with no changes that are not absolutely necessary always a good idea, makes it easier to see what's actually going on.
- Modified example is great, but for the PSE PR modifying one of the existing samples with the new API will be sufficient, otherwise a lot of duplicated sample code just to test this one thing out.

File: halo2_proofs/src/plonk/circuit.rs (hunk 6 above)
Comment (Brechtpd) [reply halo2_proofs/src/plonk/circuit.rs:834]: Or alternatively, modify `query_fixed` etc... to not set the query index anymore either. That's probably better because then the Expressions always have the same state no matter what API the user used when they get passed into halo2, so there's no 2 possible scenario's anymore. This also means we don't _really_ need additional tests anymore (except just using the new API once somewhere) as all methods of querying are basically the same.

Comment (CeciliaZ030) [reply halo2_proofs/src/plonk/circuit.rs:834]: So I'll leave it the same in this recursive method and change the cells.query_X().
Yea you're right, with this PR merged in there will be incompatibility where ConstraintSystem.advice_queries.push((column, at)) will be called twice.

Comment (CeciliaZ030) [reply halo2_proofs/src/plonk/circuit.rs:834]: So it's better to change the old API so that we only push the queries at the end of create_gate

File: halo2_proofs/src/plonk/circuit.rs (hunk 8 above)
Comment (CeciliaZ030) [reply halo2_proofs/src/plonk/circuit.rs:1878]: Yes, because `cells.queried_selectors `and `queried_cells` do not got filled in before `load_to_virtual_cells` is called

File: halo2_proofs/src/plonk/circuit.rs (hunk 6 above)
Comment (Brechtpd) [reply halo2_proofs/src/plonk/circuit.rs:834]: Yeah I think so

File: halo2_proofs/src/plonk/circuit.rs [hunk 9]
```diff
@@ -532,6 +610,14 @@ pub trait Assignment<F: Field> {
...
         N: FnOnce() -> NR;
 
+    // /// Allows the developer to include an annotation for an specific column within a `Region`.
+    // ///
+    // /// This is usually useful for debugging circuit failures.
+    // fn annotate_column<A, AR>(&mut self, annotation: A, column: Column<Any>)
+    //     where
+    //         A: FnOnce() -> AR,
+    //         AR: Into<String>;
+
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:620]: TODO?

File: halo2_proofs/src/plonk/circuit.rs [hunk 10]
```diff
@@ -1537,17 +1668,16 @@ impl<F: Field> ConstraintSystem<F> {
         let mut cells = VirtualCells::new(self);
         let table_map = table_map(&mut cells)
             .into_iter()
-            .map(|(input, table)| {
+            .map(|(mut input, mut table)| {
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:1671]:
```suggestion
            .map(|(mut input, table)| {
```

File: halo2_proofs/src/plonk/circuit.rs [hunk 11]
```diff
@@ -234,10 +256,56 @@ impl PartialOrd for Any {
...
+    }
+}
+impl ColumnType for Any {
+    fn query_cell<F: Field>(&self, index: usize, at: Rotation) -> Expression<F> {
+        match self {
+            Any::Advice(Advice { phase }) => Expression::Advice(AdviceQuery {
+                index: None,
+                column_index: index,
+                rotation: at,
+                phase: phase.clone(),
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:294]:
```suggestion
                phase: *phase,
```

**Review by Brechtpd — COMMENTED** [review]
Nice, cleaner diff! Some small suggestions, and don't forget to revert the changes to `Cargo.toml`, `primitives.rs` and remove `simple-example-2.rs`. Then I think we're ready to make a PR on the PSE repo to get some feedback on this approach.

File: halo2_proofs/src/plonk/circuit.rs [hunk 12]
```diff
@@ -706,6 +792,51 @@ pub enum Expression<F> {
 }
 
 impl<F: Field> Expression<F> {
+    /// Make side effects
+    pub fn query_cells(&mut self, cells: &mut VirtualCells<'_, F>) {
```
Comment (CPerezz) [inline halo2_proofs/src/plonk/circuit.rs:796]: This should be decently parallelizable no? Can't you split all the `cells` into chunks and digest them in parallel? Maybe `Mutex`s add too much overhead and it isn't worth it. But might be worth a try in the future probably.

Comment (CeciliaZ030) [reply halo2_proofs/src/plonk/circuit.rs:796]: This is just to make the query API easier to use, but it is a good idea to parallelize when processing `Vec<Expression<F>>` in `Gate`, or even aggregate all expressions in `ConstraintSystem` and parallelize them all. We can open an issue on that.

File: halo2_proofs/src/plonk/circuit.rs (hunk 9 above)
Comment (CPerezz) [reply halo2_proofs/src/plonk/circuit.rs:620]: https://github.com/zcash/halo2/pull/706 is ready now.  This should be closer to what we will have in the future.
